### ISSUE TITLE
Avoid writing the metrics file during the tests

### ISSUE THREAD

**Issue by kittaakos** [crapo/sadlos2#241]
Does not cause any problem, but still generates a lot of noise when running the tests:
```
java.io.FileNotFoundException: /Users/akos.kitta/OwlModels/MyModel.sadl.metrics.owl (No such file or directory)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:101)
	at com.ge.research.sadl.jena.MetricsProcessor.saveMetrics(MetricsProcessor.java:116)
	at com.ge.research.sadl.jena.JenaBasedSadlModelProcessor.onValidate(JenaBasedSadlModelProcessor.java:1099)
	at com.ge.research.sadl.processing.CompositeModelProcessor.lambda$0(CompositeModelProcessor.java:47)
	at java.lang.Iterable.forEach(Iterable.java:75)
	at com.ge.research.sadl.processing.CompositeModelProcessor.onValidate(CompositeModelProcessor.java:49)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:69)
	at org.eclipse.xtext.validation.ResourceValidatorImpl.validate(ResourceValidatorImpl.java:90)
	at com.ge.research.sadl.validation.ResourceValidator.lambda$0(ResourceValidator.java:54)
	at org.eclipse.xtext.util.OnChangeEvictingCache.get(OnChangeEvictingCache.java:77)
	at com.ge.research.sadl.validation.ResourceValidator.validate(ResourceValidator.java:56)
	at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.validate(IncrementalBuilder.java:235)
	at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.lambda$launch$6(IncrementalBuilder.java:213)
	at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.lambda$executeClustered$1(ClusteringStorageAwareResourceLoader.java:77)
	at org.eclipse.xtext.xbase.lib.internal.FunctionDelegate.apply(FunctionDelegate.java:42)
	at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:572)
	at java.util.AbstractList$Itr.next(AbstractList.java:358)
	at java.util.AbstractCollection.toArray(AbstractCollection.java:141)
	at java.util.ArrayList.addAll(ArrayList.java:577)
	at com.google.common.collect.Iterables.addAll(Iterables.java:349)
	at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.executeClustered(ClusteringStorageAwareResourceLoader.java:80)
	at org.eclipse.xtext.build.BuildContext.executeClustered(BuildContext.java:55)
	at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.launch(IncrementalBuilder.java:220)
	at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:368)
	at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:353)
	at org.eclipse.xtext.ide.server.ProjectManager.doBuild(ProjectManager.java:113)
	at org.eclipse.xtext.ide.server.BuildManager.internalBuild(BuildManager.java:169)
	at org.eclipse.xtext.ide.server.BuildManager.doBuild(BuildManager.java:125)
	at org.eclipse.xtext.ide.server.WorkspaceManager.doBuild(WorkspaceManager.java:153)
	at org.eclipse.xtext.ide.server.WorkspaceManager.didOpen(WorkspaceManager.java:202)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$didOpen$4(LanguageServerImpl.java:249)
	at org.eclipse.xtext.testing.AbstractLanguageServerTest$1$1.runWrite(AbstractLanguageServerTest.java:127)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.didOpen(LanguageServerImpl.java:252)
	at org.eclipse.xtext.testing.AbstractLanguageServerTest.open(AbstractLanguageServerTest.java:235)
	at org.eclipse.xtext.testing.AbstractLanguageServerTest.open(AbstractLanguageServerTest.java:218)
	at com.ge.research.sadl.tests.lsp.SadlColoringTest.checkColoring_03(SadlColoringTest.java:166)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:86)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:38)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:459)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:678)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:382)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:192)
```

**Comment by crapo** [comment]
what is the best way to detect that this is a test and therefore not do metrics processing?

**Comment by crapo** [comment]
Humm... I'm a little confused. The call to saveMetrics in JenaBasedSadlModelProcessor.onValidate, line 1099, is wrapped in this test:
```
					if (!isSyntheticUri(null, resource)) {
                                                ...
                                        }
```
How do you run a test case without the URI being synthetic?

**Comment by kittaakos** [comment]
I will look into this.

**Comment by kittaakos** [comment]
We use real, file EMF URIs for the tests we have established for the web version of SADL. I will try to switch to synthetic URIs, if it does not work, I will add a second guard before trying to write the metrics file.

**Comment by kittaakos** [comment]
We would have to switch to the latest Xtext + LSP versions. With the current LSP version, we need to have file URI scheme. In memory workspace with custom URI schemes (such as `synthetic://`) is not supported with our current version. I will add a guard to the metrics processor.

**Comment by kittaakos** [comment]
Once we switch to the latest Xtext + LSP, we could do something like this in our tests:
```diff
diff --git a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/helpers/SadlTestProjectStructureInitializer.xtend b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/helpers/SadlTestProjectStructureInitializer.xtend
index ee09cfd..91f8fc9 100644
--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/helpers/SadlTestProjectStructureInitializer.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/helpers/SadlTestProjectStructureInitializer.xtend
@@ -86,9 +86,6 @@
 	Test: area of MyCircle is 63.61.
 	
 	Test: area of MyCircle is 53.51.''';
-
-	@Inject
-	DotProjectContentProvider dotProjectContentProvider;
 	
 	@Inject
 	ISadlImplicitModelContentProvider modelContentProvider;
@@ -99,65 +96,36 @@
 	 * 
 	 * @param projectRoot the path to the project root.
 	 */
-	def void initialize(Path projectRoot) {
-		// This is just a hack to be able to locate the project root and 
-		// the implicit model folder in a web project as well. 
-		projectRoot.createDotProjectFileIfMissing;
-
-		projectRoot.createShapesFileIfMissing;
-		projectRoot.createCircleFileIfMissing;
-		projectRoot.createRectangleFileIfMissing;
-		projectRoot.createTestFileIfMissing;
-
-		val implicitModelRoot = projectRoot.resolve(SADL_IMPLICIT_MODEL_FOLDER);
-		if (!implicitModelRoot.toFile.exists) {
-			Files.createDirectory(implicitModelRoot);
-		}
-		implicitModelRoot.createImplicitModelFileIfMissing;
-
-		val owlModelsRoot = projectRoot.resolve(OWL_MODELS_FOLDER_NAME);
-		if (!owlModelsRoot.toFile.exists) {
-			Files.createDirectories(owlModelsRoot);
-		}
+	def void initialize((String, String)=> void resourceInitializer) {
+		resourceInitializer.createShapesFileIfMissing;
+		resourceInitializer.createCircleFileIfMissing;
+		resourceInitializer.createRectangleFileIfMissing;
+		resourceInitializer.createTestFileIfMissing;
+		resourceInitializer.createImplicitModel;
 	}
 
-	private def void createDotProjectFileIfMissing(Path projectRoot) {
-		val projectName = projectRoot.toFile.name;
-		projectRoot.createFileIfMissing(DOT_PROJECT_FILENAME, dotProjectContentProvider.getContent(projectName));
+	private def void createImplicitModel((String, String)=> void resourceInitializer) {
+		resourceInitializer.apply(SADL_IMPLICIT_MODEL_FILENAME.toUri, modelContentProvider.content);
+	}
+
+	private def void createShapesFileIfMissing((String, String)=> void resourceInitializer) {
+		resourceInitializer.apply(SHAPES_FILENAME.toUri, SHAPES_FILE_CONTENT);
 	}
 	
-	private def void createShapesFileIfMissing(Path projectRoot) {
-		projectRoot.createFileIfMissing(SHAPES_FILENAME, SHAPES_FILE_CONTENT);
+	private def void createCircleFileIfMissing((String, String)=> void resourceInitializer) {
+		resourceInitializer.apply(CIRCLE_FILENAME.toUri, CIRCLE_FILE_CONTENT);
+	}
+
+	private def void createRectangleFileIfMissing((String, String)=> void resourceInitializer) {
+		resourceInitializer.apply(RECTANGLE_FILENAME.toUri, RECTANGLE_FILE_CONTENT);
 	}
 	
-	private def void createCircleFileIfMissing(Path projectRoot) {
-		projectRoot.createFileIfMissing(CIRCLE_FILENAME, CIRCLE_FILE_CONTENT);
-	}
-
-	private def void createRectangleFileIfMissing(Path projectRoot) {
-		projectRoot.createFileIfMissing(RECTANGLE_FILENAME, RECTANGLE_FILE_CONTENT);
+	private def void createTestFileIfMissing((String, String)=> void resourceInitializer) {
+		resourceInitializer.apply(TEST_FILENAME.toUri, TEST_FILE_CONTENT);	
 	}
 	
-	private def void createTestFileIfMissing(Path projectRoot) {
-		projectRoot.createFileIfMissing(TEST_FILENAME, TEST_FILE_CONTENT);	
-	}
-
-	private def createImplicitModelFileIfMissing(Path implicitModelRoot) {
-		implicitModelRoot.createFileIfMissing(SADL_IMPLICIT_MODEL_FILENAME, modelContentProvider.content);
-	}
-
-	private def createFileIfMissing(Path rootFolder, String fileName, String content) {
-		rootFolder.toFile.directory.
-			checkArgument('''Cannot find root directory for file «fileName» under: «rootFolder».''');
-		val filePath = rootFolder.resolve(fileName);
-		val file = filePath.toFile;
-		if (!file.exists) {
-			Files.createFile(filePath);
-			val writer = new FileWriter(file);
-			writer.write(content);
-			writer.flush();
-			writer.close();
-		}
+	private def toUri(String fileName) {
+		return '''synthetic://test/«fileName»''';
 	}
 
 }
diff --git a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/lsp/SadlColoringTest.xtend b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/lsp/SadlColoringTest.xtend
index 595b444..81babf3 100644
--- a/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/lsp/SadlColoringTest.xtend
+++ b/sadl3/com.ge.research.sadl.parent/com.ge.research.sadl.tests/src/com/ge/research/sadl/tests/lsp/SadlColoringTest.xtend
@@ -123,8 +123,8 @@
 	}
 
 	override protected initialize() {
-		val rootPath = root.absoluteFile.toPath;
-		projectStructureInitializer.initialize(rootPath);
+		val (String, String)=> void initializer = [ uri, content | open(uri, content)];
+		projectStructureInitializer.initialize(initializer);
 		super.initialize();
 	}
```

**Comment by kittaakos** [comment]
PR: https://github.com/crapo/sadlos2/pull/258

**Comment by kittaakos** [comment]
From: https://github.com/crapo/sadlos2/issues/265#issue-266636492

**Comment by kittaakos** [comment]
This will be fixed as part of #292.
